### PR TITLE
Merged phase_1 and phase_2 into train.py

### PR DESCRIPTION
**Pull request by captain-pool**
Addresses Issue: #30 
Merged Phase 1 and Phase 2 trainer.
Tn this PR the two functions from those different files(phase_1.py and phase_2.py) are merged into one file called `train.py`. Since these are two independent functions, some of the same variables are being reused and reloaded in each of the cases, which is not one of the finest solution, provided it may take some time, onloading and offloading the dataset from the memory.

Proposed Design Suggestion
-----------
Instead of using different functions and using similar variables in two different functions,
It would be better to use this design, saving memory and time wasted in on loading and off loading data, (especially, reloading the entire dataset) to and from the memory.
```python3
class Trainer(object):
  @classmethod
  def setup_training(cls, Generator, Discriminator, summary_writer, data_dir=None):
    cls.G = Generator
    cls.D = Discriminator
    cls.summary_writer = summary_writer
    cls.dataset = utils.load_dataset(data_dir=data_dir, ...)
  @classmethod
  def train_psnr(cls):
    # use class params:
    # cls.G, cls.D, cls.summary_writer, cls.dataset
    # to train the psnr model
    pass
  @classmethod
  def train_gan(cls):
    # use class params:
    # cls.G, cls.D, cls.summary_writer, cls.dataset
    # to train ESRGAN model
    pass
```
@srjoglekar246 what do you think?

### CODE REVIEW

File: E2_ESRGAN/lib/train.py [hunk 1]
```diff
@@ -3,25 +3,102 @@
...
+        logging.info(
+            "[WARMUP] Epoch: %d\tBatch: %d\tGenerator Loss: %f\tTime Taken: %f" %
+            (epoch, step // (epoch + 1), mean_loss.numpy(), time.time() - start_time))
+        if mean_loss < previous_loss:
+          utils.save_checkpoint(checkpoint, "phase_1")
+        previous_loss = mean_loss
+        start_time = time.time()
+
+
+def train_gan(G, D, summary_writer, settings, data_dir=None):
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: Google internal style guide usually discourages single-letter variable names :-)

Comment (captain-pool) [reply E2_ESRGAN/lib/train.py]: Is that guide accessible? 😅 I can then take a look and work accordingly.

Comment (srjoglekar246) [reply E2_ESRGAN/lib/train.py]: http://google.github.io/styleguide/pyguide.html#316-naming

File: E2_ESRGAN/lib/train.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+import time
+import logging
+import itertools
+from functools import partial
+import tensorflow as tf
+from lib import utils, model, dataset
+
+
+class Training(object):
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: docstrings :-). Doesn't need much, just the utility of this class.

File: E2_ESRGAN/lib/train.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  @classmethod
+  def setup_training(cls, summary_writer, settings, data_dir=None):
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: I think its better to just make this instance-specific. That way, a client could use this trainer to train multiple models in parallel without possibility of side-effects.

File: E2_ESRGAN/lib/train.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    dataset_args = cls.settings["dataset"]
+    cls.dataset = dataset.load_dataset(
+        dataset_args["name"],
+        dataset.scale_down(
+            method=dataset_args["scale_method"],
+            dimension=dataset_args["hr_dimension"]),
+        batch_size=settings["batch_size"],
+        data_dir=data_dir)
+
+  @classmethod
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: Same for these methods. We can have them be non-static.

File: E2_ESRGAN/lib/train.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    perceptual_loss = utils.PerceptualLoss(
+        weights="imagenet",
+        input_shape=[
+            cls.settings["dataset"]["hr_dimension"],
+            cls.settings["dataset"]["hr_dimension"]])
+    Ra_G = utils.RelativisticAverageLoss(discriminator, type_="G")
+    Ra_D = utils.RelativisticAverageLoss(discriminator, type_="D")
+
+    hot_start = tf.train.Checkpoint(G=generator, G_optimizer=G_optimizer)
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: What is hot_start used for? Please document.

**Comment by captain-pool** [comment]
After this gets merged, I will add distribution strategy in a new PR.
Thus addressing Issue #29

File: E2_ESRGAN/lib/train.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
+import time
+import logging
+import itertools
+from functools import partial
+import tensorflow as tf
+from lib import utils, model, dataset
+
+
+class Training(object):
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/train.py]: nit: Can we name this `Trainer`?

Comment (captain-pool) [reply E2_ESRGAN/lib/train.py]: Sure! :)

**Pull request merged**